### PR TITLE
consider aliases and path names when looking for dupe imports

### PR DESCRIPTION
**Pull request by sjamesr**
Addresses https://github.com/go-lang-plugin-org/go-lang-idea-plugin/issues/1461

### CODE REVIEW

**Comment by ignatov** [comment]
Thanks

**Comment by ignatov** [comment]
As for contribution agreement, please download the [PDF](http://www.jetbrains.org/download/attachments/983565/JB_ContributorLicenseAgreement.pdf?version=3&modificationDate=1416937277000) and then print it out, sign (mention you github account name there), scan and email to contribute@jetbrains.com

**Comment by sjamesr** [comment]
Hey Sergey,

Jetbrains should already have a CLA on file for me (sent September 11,
2014). Please let me know if you need another.

Regards,
James

File: src/com/goide/inspections/GoUnusedImportDeclaration.java [hunk 1]
```diff
@@ -85,4 +105,25 @@ public void visitElement(@NotNull PsiElement o) {
       }
     });
   }
+
+  /**
+   * Returns a map of import 'name' to import spec, where name is either the import's alias or the last component of the import's path. The
+   * returned map orders imports the same way as they occur in the file.
+   */
+  private static MultiMap<String, GoImportSpec> getImportNameMap(@NotNull GoFile file) {
```
Comment (zolotov) [inline src/com/goide/inspections/GoUnusedImportDeclaration.java]: What's the difference with GoFile#getImportMap method?

File: src/com/goide/inspections/GoUnusedImportDeclaration.java [hunk 2]
```diff
@@ -62,6 +69,19 @@ protected void checkFile(@NotNull GoFile file, @NotNull ProblemsHolder problemsH
     for (GoImportSpec duplicatedImportSpec : GoImportOptimizer.findDuplicatedEntries(importMap)) {
       problemsHolder.registerProblem(duplicatedImportSpec, "Redeclared import", ProblemHighlightType.GENERIC_ERROR, OPTIMIZE_QUICK_FIX);
     }
+
+    // Find imports that are referred to by the same alias or last pathname component.
+    MultiMap<String, GoImportSpec> nameMap = getImportNameMap(file);
```
Comment (zolotov) [inline src/com/goide/inspections/GoUnusedImportDeclaration.java]: This logic partly duplicates the loop above that is using GoImportOptimizer#findDuplicatedEntries. This leads to duplicated error message
screenshot 2015-09-09 21 44 35

File: src/com/goide/inspections/GoUnusedImportDeclaration.java [hunk 3]
```diff
@@ -62,6 +69,19 @@ protected void checkFile(@NotNull GoFile file, @NotNull ProblemsHolder problemsH
...
+    for (Map.Entry<String, Collection<GoImportSpec>> specs : nameMap.entrySet()) {
+      // These iterators are guaranteed to be non-empty.
+      Iterator<GoImportSpec> imports = specs.getValue().iterator();
+      // Add warnings to the second and subsequent imports.
+      imports.next();
+      while (imports.hasNext()) {
+        problemsHolder.registerProblem(imports.next(), "Redeclared import", ProblemHighlightType.GENERIC_ERROR, OPTIMIZE_QUICK_FIX);
```
Comment (zolotov) [inline src/com/goide/inspections/GoUnusedImportDeclaration.java]: OptimizeImports quick fix does nothing with partially matched imports so this problem shouldn't declare it and possible fix.

File: src/com/goide/inspections/GoUnusedImportDeclaration.java (hunk 2 above)
Comment (sjamesr) [inline src/com/goide/inspections/GoUnusedImportDeclaration.java]: Should be fixed, now I only suggest a fix if the import wasn't already flagged in the above loop.

File: src/com/goide/inspections/GoUnusedImportDeclaration.java (hunk 1 above)
Comment (sjamesr) [inline src/com/goide/inspections/GoUnusedImportDeclaration.java]: Removed, it caused a redeclared import to be detected in another test, can you make sure the test case is still correct?

File: src/com/goide/inspections/GoUnusedImportDeclaration.java (hunk 3 above)
Comment (sjamesr) [inline src/com/goide/inspections/GoUnusedImportDeclaration.java]: Done, thanks

**Comment by sjamesr** [comment]
Thank you for the review, the commit should be in better shape now

File: src/com/goide/inspections/GoUnusedImportDeclaration.java (hunk 1 above)
Comment (zolotov) [inline src/com/goide/inspections/GoUnusedImportDeclaration.java]: Sure, I'll check

File: src/com/goide/inspections/GoUnusedImportDeclaration.java (hunk 2 above)
Comment (zolotov) [inline src/com/goide/inspections/GoUnusedImportDeclaration.java]: Looks like it most consistent solution

**Pull request merged**